Seg fault running FileSystemWatcher tests on OSX

http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/outerloop_osx_debug/52/consoleText
```
/Users/dotnet-bot/j/workspace/dotnet_corefx/master/outerloop_osx_debug/bin/tests/OSX.AnyCPU.Debug/System.IO.FileSystem.Watcher.Tests/netcoreapp1.0/RunTests.sh: line 285: 70576 Segmentation fault: 11  ./corerun xunit.console.netcore.exe System.IO.FileSystem.Watcher.Tests.dll -xml testResults.xml -notrait Benchmark=true -notrait category=IgnoreForCI -notrait category=failing -notrait category=nonosxtests
  Finished running tests. End time=23:20:31. Return value was 139
/Users/dotnet-bot/j/workspace/dotnet_corefx/master/outerloop_osx_debug/Tools/tests.targets(202,5): warning MSB3073: The command "/Users/dotnet-bot/j/workspace/dotnet_corefx/master/outerloop_osx_debug/bin/tests/OSX.AnyCPU.Debug/System.IO.FileSystem.Watcher.Tests/netcoreapp1.0/RunTests.sh /Users/dotnet-bot/j/workspace/dotnet_corefx/master/outerloop_osx_debug/packages/" exited with code 139. [/Users/dotnet-bot/j/workspace/dotnet_corefx/master/outerloop_osx_debug/src/System.IO.FileSystem.Watcher/tests/System.IO.FileSystem.Watcher.Tests.csproj]
/Users/dotnet-bot/j/workspace/dotnet_corefx/master/outerloop_osx_debug/Tools/tests.targets(211,5): error : One or more tests failed while running tests from 'System.IO.FileSystem.Watcher.Tests' please check log for details! [/Users/dotnet-bot/j/workspace/dotnet_corefx/master/outerloop_osx_debug/src/System.IO.FileSystem.Watcher/tests/System.IO.FileSystem.Watcher.Tests.csproj]
```
cc: @janvorli, @ianhays

--- a/src/Common/src/Interop/OSX/Interop.CoreFoundation.cs
+++ b/src/Common/src/Interop/OSX/Interop.CoreFoundation.cs
@@ -89,6 +89,16 @@ internal static partial class Interop
         }
 
         /// <summary>
+        /// You should retain a Core Foundation object when you receive it from elsewhere
+        /// (that is, you did not create or copy it) and you want it to persist. If you 
+        /// retain a Core Foundation object you are responsible for releasing it
+        /// </summary>
+        /// <param name="ptr">The CFType object to retain. This value must not be NULL</param>
+        /// <returns>The input value</param>
+        [DllImport(Interop.Libraries.CoreFoundationLibrary)]
+        internal extern static IntPtr CFRetain(IntPtr ptr);
+
+        /// <summary>
         /// Decrements the reference count on the specified object and, if the ref count hits 0, cleans up the object.
         /// </summary>
         /// <param name="ptr">The pointer on which to decrement the reference count.</param>

--- a/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs
+++ b/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs
@@ -140,6 +140,7 @@ namespace System.IO
 
             // Calling RunLoopStop multiple times SegFaults so protect the call to it
             private bool _stopping;
+            private object StopLock => this;
 
             internal RunningInstance(
                 FileSystemWatcher watcher,
@@ -152,6 +153,7 @@ namespace System.IO
                 Debug.Assert(!cancelToken.IsCancellationRequested);
 
                 _weakWatcher = new WeakReference<FileSystemWatcher>(watcher);
+                _watcherRunLoop = IntPtr.Zero;
                 _fullDirectory = System.IO.Path.GetFullPath(directory);
                 _includeChildren = includeChildren;
                 _filterFlags = filter;
@@ -162,12 +164,15 @@ namespace System.IO
 
             private void CancellationCallback()
             {
-                if (!_stopping)
+                lock (StopLock)
                 {
-                    _stopping = true;
+                    if (!_stopping && _watcherRunLoop != IntPtr.Zero)
+                    {
+                        _stopping = true;
 
-                    // Stop the FS event message pump
-                    Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
+                        // Stop the FS event message pump
+                        Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
+                    }
                 }
             }
 
@@ -244,6 +249,10 @@ namespace System.IO
                 _watcherRunLoop = Interop.RunLoop.CFRunLoopGetCurrent();
                 Debug.Assert(_watcherRunLoop != IntPtr.Zero);
 
+                // Retain the RunLoop so that it doesn't get moved or cleaned up before we're done with it.
+                IntPtr retainResult = Interop.CoreFoundation.CFRetain(_watcherRunLoop);
+                Debug.Assert(retainResult == _watcherRunLoop, "CFRetain is supposed to return the input value");
+
                 // Schedule the EventStream to run on the thread's RunLoop
                 Interop.EventStream.FSEventStreamScheduleWithRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
 
@@ -278,6 +287,13 @@ namespace System.IO
                 {
                     // Always unschedule the RunLoop before cleaning up
                     Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+
+                    // Release the WatcherLoop Core Foundation object.
+                    lock (StopLock)
+                    {
+                        Interop.CoreFoundation.CFRelease(_watcherRunLoop);
+                        _watcherRunLoop = IntPtr.Zero;
+                    }
                 }
             }
 

